File: src/android/CDVDistimoSDK.java
package com.distimo.sdk.cordova;

import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.PluginResult;
import org.json.JSONArray;
import org.json.JSONException;

import com.distimo.sdk.DistimoSDK;

public class CDVDistimoSDK extends CordovaPlugin {

	@Override
	public boolean execute(String action, final JSONArray args, final CallbackContext callbackContext) throws JSONException {
		try {
			if (action.equals("start")) {
				if (args.length() >= 1) {
					String sdkKey = args.getString(0);
					this.start(sdkKey, callbackContext);
					return true;
				}
			} else if (action.equals("version")) {
				this.version(callbackContext);
				return true;
			} else if (action.equals("logUserRegistered")) {
				this.logUserRegistered(callbackContext);
				return true;
			} else if (action.equals("logInAppPurchase")) {
				if (args.length() >= 2) {
					String productID = args.getString(0);
					String orderID = args.getString(1);
					this.logInAppPurchase(productID, orderID, callbackContext);
					return true;
				}
			} else if (action.equals("logInAppPurchaseRefunded")) {
				if (args.length() >= 2) {
					String productID = args.getString(0);
					String orderID = args.getString(1);
					this.logInAppPurchaseRefunded(productID, orderID, callbackContext);
					return true;
				}
			} else if (action.equals("logExternalPurchaseWithCurrency")) {
				if (args.length() >= 4) {
					String productID = args.getString(0);
					String currency = args.getString(1);
					double price = args.getDouble(2);
					int quantity = args.getInt(3);
					this.logExternalPurchaseWithCurrency(productID, currency, price, quantity, callbackContext);
					return true;
				}
			} else if (action.equals("logBannerClick")) {
				String publisher = (args.isNull(0) ? null : args.getString(0));
				this.logBannerClick(publisher, callbackContext);
				return true;
			} else if (action.equals("setUserID")) {
				String userID = (args.isNull(0) ? null : args.getString(0));
				this.setUserID(userID, callbackContext);
				return true;
			} else if (action.equals("openAppLink")) {
				if (args.length() >= 1) {
					String applinkHandle = args.getString(0);
					String campaignHandle = (args.isNull(1) ? null : args.getString(1));
					this.openAppLink(applinkHandle, campaignHandle, callbackContext);
					return true;
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		return false;
	}
	
	private void start(String sdkKey, CallbackContext callbackContext) {
		if (sdkKey != null && sdkKey.length() > 0) {
			DistimoSDK.onCreate(cordova.getActivity(), sdkKey);
			callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
		} else {
			callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR, "Please provide a valid SDK Key, you can create one at https://analytics.distimo.com/settings/sdk."));
		}
	}
	
	private void version(CallbackContext callbackContext) {
		String version = DistimoSDK.version();
		if (version != null && version.length() > 0) {
			callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
		} else {
			callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.ERROR, "No DistimoSDK version found."));
		}
	}
	
	private void logUserRegistered(CallbackContext callbackContext) {
		DistimoSDK.onUserRegistered();
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void logInAppPurchase(String productID, String orderID, CallbackContext callbackContext) {
		DistimoSDK.onInAppPurchase(productID, orderID);
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void logInAppPurchaseRefunded(String productID, String orderID, CallbackContext callbackContext) {
		DistimoSDK.onInAppPurchaseRefunded(productID, orderID);
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void logExternalPurchaseWithCurrency(String productID, String currency, double price, int quantity, CallbackContext callbackContext) {
		DistimoSDK.onExternalPurchase(productID, currency, price, quantity);
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void logBannerClick(String publisher, CallbackContext callbackContext) {
		DistimoSDK.onBannerClick(publisher);
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void setUserID(String userID, CallbackContext callbackContext) {
		DistimoSDK.setUserID(userID);
		callbackContext.sendPluginResult(new PluginResult(PluginResult.Status.OK));
	}
	
	private void openAppLink(String applinkHandle, String campaignHandle, CallbackContext callbackContext) {
		DistimoSDK.openAppLink(applinkHandle, campaignHandle, cordova.getActivity());
	}
}

File: src/ios/CDVDistimoSDK.m
//
//  CDVDistimoSDK.m
//  Analytics
//
//  Created by Arne de Vries on 4/4/13.
//
//

#import <Cordova/CDV.h>
#import "DistimoSDK.h"

@interface CDVDistimoSDK : CDVPlugin

@end

@implementation CDVDistimoSDK

#pragma mark - Start

- (void)start:(CDVInvokedUrlCommand *)command {
	NSString *sdkKey = [command.arguments objectAtIndex:0];
	
	if (sdkKey) {
		[DistimoSDK handleLaunchWithOptions:nil
									 sdkKey:sdkKey];
		
		[self callback:CDVCommandStatus_OK
			   message:nil
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Please provide a valid SDK Key, you can create one at https://analytics.distimo.com/settings/sdk."
			   command:command];
	}
}

#pragma mark - Settings

- (void)version:(CDVInvokedUrlCommand *)command {
	NSString *version = [DistimoSDK version];
	
	if (version) {
		[self callback:CDVCommandStatus_OK
			   message:version
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Could not retrieve DistimoSDK version"
			   command:command];
	}
}

#pragma mark - User Value

- (void)logUserRegistered:(CDVInvokedUrlCommand *)command {
	[DistimoSDK logUserRegistered];
	
	[self callback:CDVCommandStatus_OK
		   message:nil
		   command:command];
}

- (void)logInAppPurchaseWithLocale:(CDVInvokedUrlCommand *)command {
	if ([command.arguments count] >= 4) {
		NSString *productID = [command.arguments objectAtIndex:0];
		NSLocale *locale = [[[NSLocale alloc] initWithLocaleIdentifier:[command.arguments objectAtIndex:1]] autorelease];
		double price = [[command.arguments objectAtIndex:2] doubleValue];
		int quantity = [[command.arguments objectAtIndex:3] intValue];
		
		[DistimoSDK logInAppPurchaseWithProductID:productID
									  priceLocale:locale
											price:price
										 quantity:quantity];
		
		[self callback:CDVCommandStatus_OK
			   message:nil
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Not enough arguments provided"
			   command:command];
	}
}

- (void)logInAppPurchaseWithCurrency:(CDVInvokedUrlCommand *)command {
	if ([command.arguments count] >= 4) {
		NSString *productID = [command.arguments objectAtIndex:0];
		NSString *currency = [command.arguments objectAtIndex:1];
		double price = [[command.arguments objectAtIndex:2] doubleValue];
		int quantity = [[command.arguments objectAtIndex:3] intValue];
		
		[DistimoSDK logInAppPurchaseWithProductID:productID
									 currencyCode:currency
											price:price
										 quantity:quantity];
		
		[self callback:CDVCommandStatus_OK
			   message:nil
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Not enough arguments provided"
			   command:command];
	}
}

- (void)logExternalPurchaseWithLocale:(CDVInvokedUrlCommand *)command {
	if ([command.arguments count] >= 4) {
		NSString *productID = [command.arguments objectAtIndex:0];
		NSLocale *locale = [[[NSLocale alloc] initWithLocaleIdentifier:[command.arguments objectAtIndex:1]] autorelease];
		double price = [[command.arguments objectAtIndex:2] doubleValue];
		int quantity = [[command.arguments objectAtIndex:3] intValue];
		
		[DistimoSDK logExternalPurchaseWithProductID:productID
										 priceLocale:locale
											   price:price
											quantity:quantity];
		
		[self callback:CDVCommandStatus_OK
			   message:nil
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Not enough arguments provided"
			   command:command];
	}
}

- (void)logExternalPurchaseWithCurrency:(CDVInvokedUrlCommand *)command {
	if ([command.arguments count] >= 4) {
		NSString *productID = [command.arguments objectAtIndex:0];
		NSString *currency = [command.arguments objectAtIndex:1];
		double price = [[command.arguments objectAtIndex:2] doubleValue];
		int quantity = [[command.arguments objectAtIndex:3] intValue];
		
		[DistimoSDK logExternalPurchaseWithProductID:productID
										currencyCode:currency
											   price:price
											quantity:quantity];
		
		[self callback:CDVCommandStatus_OK
			   message:nil
			   command:command];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Not enough arguments provided"
			   command:command];
	}
}

- (void)logBannerClick:(CDVInvokedUrlCommand *)command {
	NSString *publisher = [command.arguments objectAtIndex:0];
	
	[DistimoSDK logBannerClickWithPublisher:([publisher isKindOfClass:[NSNull class]] ? nil : publisher)];
	
	[self callback:CDVCommandStatus_OK
		   message:nil
		   command:command];
}

#pragma mark - User Properties

- (void)setUserID:(CDVInvokedUrlCommand *)command {
	NSString *userID = [command.arguments objectAtIndex:0];
	
	[DistimoSDK setUserID:([userID isKindOfClass:[NSNull class]] ? nil : userID)];
	
	[self callback:CDVCommandStatus_OK
		   message:nil
		   command:command];
}

#pragma mark - AppLink

- (void)openAppLink:(CDVInvokedUrlCommand *)command {
	if ([command.arguments count] >= 1) {
		NSString *applinkHandle = [command.arguments objectAtIndex:0];
		NSString *campaignHandle = [command.arguments objectAtIndex:1];
		
		[DistimoSDK openAppLink:applinkHandle
					   campaign:([campaignHandle isKindOfClass:[NSNull class]] ? nil : campaignHandle)
			   inViewController:nil];
	} else {
		[self callback:CDVCommandStatus_ERROR
			   message:@"Not enough arguments provided"
			   command:command];
	}
	
	[self callback:CDVCommandStatus_OK
		   message:nil
		   command:command];
}

#pragma mark - Helpers

- (void)callback:(CDVCommandStatus)status message:(NSString *)message command:(CDVInvokedUrlCommand *)command {
	CDVPluginResult *result;
	
	if (message) {
		result = [CDVPluginResult resultWithStatus:status
								   messageAsString:message];
	} else {
		result = [CDVPluginResult resultWithStatus:status];
	}
	
	[self.commandDelegate sendPluginResult:result
								callbackId:command.callbackId];
}

@end
